drivers/mfd/rpi-rp2040-gpio-bridge.c

```diff
@@ -0,0 +1,1155 @@
+				     value ? pattern : 0, !value ? pattern : 0);
+	if (ret) {
+		dev_err(&client->dev, "%s(%u, %d) could not ST_CL\n", __func__,
+			offset, value);
+		return ret;
+	}
+
+	priv_data->gpio_direction &= ~pattern;
+	ret = rp2040_gbdg_10byte_cmd(client, CMD_GPIO_OE,
+				     ~priv_data->gpio_direction,
+				     priv_data->gpio_direction);
+	return ret;
+}
+
+static int rp2040_gbdg_gpio_get(struct gpio_chip *gc, unsigned int offset)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	struct i2c_client *client = priv_data->client;
+	struct rp2040_gbdg_device_info info;
+	int ret;
+
+	if (offset >= NUM_GPIO)
+		return -EINVAL;
+
+	ret = rp2040_gbdg_get_device_info(client, &info);
+	if (ret)
+		return ret;
+
+	return info.status & (1 << (offset + 8)) ? 1 : 0;
+}
+
+static void rp2040_gbdg_gpio_set(struct gpio_chip *gc, unsigned int offset,
+				 int value)
+{
+	struct rp2040_gbdg *priv_data = gpiochip_get_data(gc);
+	struct i2c_client *client = priv_data->client;
+	u32 pattern;
+	int ret;
+
+	if (offset >= NUM_GPIO) {
+		ret = -EINVAL;
+	} else {
+		pattern = (1 << (offset + 8));
+		ret = rp2040_gbdg_10byte_cmd(client, CMD_GPIO_ST_CL,
+					     value ? pattern : 0,
+					     !value ? pattern : 0);
+	}
+}
+
+static int rp2040_gbdg_probe(struct i2c_client *client)
+{
+	struct device *dev = &client->dev;
+	struct device_node *np = dev->of_node;
+	struct rp2040_gbdg *priv_data;
+	struct spi_controller *controller;
+	struct regulator *reg;
+	struct rp2040_gbdg_device_info info;
+	struct device_node *dn;
+	struct of_phandle_args of_args[2] = { 0 };
+	int ret;
+
+	reg = devm_regulator_get(dev, "power");
+	if (IS_ERR(reg))
+		return dev_err_probe(dev, PTR_ERR(reg),
+				     "error getting regulator\n");
+
+	ret = regulator_enable(reg);
+	if (ret)
+		return dev_err_probe(dev, ret, "could not enable regulator\n");
+
+	ret = rp2040_gbdg_get_device_info(client, &info);
+	if (ret) {
+		regulator_disable(reg);
+		return dev_err_probe(dev, ret, "error getting device info\n");
+	}
+	dev_info(dev, "%s() found dev ID: %llx, fw ver. %u\n", __func__,
+		 info.id, info.version);
+
+	ret = regulator_disable(reg);
+	if (ret)
+		return dev_err_probe(dev, ret, "could not disable regulator\n");
+
+	devm_pm_runtime_enable(dev);
+	pm_runtime_use_autosuspend(dev);
+
+	controller = devm_spi_alloc_master(dev, sizeof(struct rp2040_gbdg));
+	if (!controller)
+		return dev_err_probe(dev, ENOMEM,
+				     "could not alloc spi controller\n");
+
+	priv_data = spi_controller_get_devdata(controller);
+	priv_data->controller = controller;
+	priv_data->client = client;
+	priv_data->regulator = reg;
+
+	i2c_set_clientdata(client, priv_data);
+
+	priv_data->shash = crypto_alloc_shash("md5", 0, 0);
+	if (IS_ERR(priv_data->shash))
+		return dev_err_probe(dev, PTR_ERR(priv_data->shash),
+				     "Could not allocate shash\n");
+
+	if (crypto_shash_digestsize(priv_data->shash) != MD5_DIGEST_SIZE) {
+		crypto_free_shash(priv_data->shash);
```

Comment:
Repeated multiple times. It might be nice to make it a common exit path at the end of the function.